File: packages/typescript/goldenmatch/src/core/identity/audit.ts
/**
 * audit.ts -- byte-identical TS port of Python `identity/audit.py` (#1078).
 *
 * Tamper-evidence for the append-only identity event log, in two layers:
 *   1. per-event content hash (`eventContentHash`) computed at insert time;
 *   2. an on-demand seal chain (`sealAuditLog` / `verifyAuditChain`) that folds
 *      every event's content hash, in `eventId` order, into a chained root.
 *
 * The whole point of this port is CROSS-VERIFICATION: a seal (or entry hash)
 * computed here must be BYTE-IDENTICAL to Python's, so a `.goldenmatch/
 * identity.db` written by either toolkit verifies under the other. That hinges
 * on reproducing Python's `json.dumps(canon, sort_keys=True,
 * separators=(",",":"))` (default `ensure_ascii=True`) EXACTLY -- see
 * `canonicalJson` below. Do NOT swap in `JSON.stringify`; it diverges on key
 * ordering, non-ASCII escaping, and float repr (`1.0` vs `1`).
 *
 * Edge-safe: uses Web Crypto (`crypto.subtle`) + `TextEncoder`, no `node:*`.
 * Hashing is async (not a hot path).
 */
import type { IdentityEvent, AuditSeal, IdentityStore } from "./types.js";
import { pyIsoformat } from "./pyDatetime.js";

/**
 * Wrapper marking a JS number that must serialize as a PYTHON FLOAT (trailing
 * `.0` when integral), matching how Python's `json.dumps` renders a `float`.
 * JS collapses `1.0` to the integer `1`; Python keeps `1.0`. `trust` is the one
 * canon field that is a Python float, so it is wrapped; `previousClaimId` is a
 * Python `int` and stays a plain number (integer repr).
 */
class PyFloat {
  constructor(readonly value: number) {}
}

/** Reproduce Python `round(x, 6)` for the trust field (values are 0..1, ≤6dp
 * in practice, so this is effectively identity; kept for faithfulness). */
function round6(x: number): number {
  return Math.round(x * 1e6) / 1e6;
}

/** Escape a string the way Python `json.dumps(..., ensure_ascii=True)` does:
 * `"`/`\` escaped, the five short control escapes, every other codepoint
 * `< 0x20` or `> 0x7E` as lowercase `\uXXXX` (astral chars as their UTF-16
 * surrogate halves, exactly as Python emits them). */
function encodeString(s: string): string {
  let out = '"';
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i);
    if (c === 0x22) out += '\\"';
    else if (c === 0x5c) out += "\\\\";
    else if (c === 0x08) out += "\\b";
    else if (c === 0x09) out += "\\t";
    else if (c === 0x0a) out += "\\n";
    else if (c === 0x0c) out += "\\f";
    else if (c === 0x0d) out += "\\r";
    else if (c < 0x20 || c > 0x7e) out += "\\u" + c.toString(16).padStart(4, "0");
    else out += s[i];
  }
  return out + '"';
}

/** Render a JS number as Python's `json.dumps` would render the corresponding
 * Python value. Integral -> integer repr (`5`); otherwise JS shortest repr,
 * which matches Python's float repr for the values that occur here. */
function encodeNumber(n: number): string {
  if (!Number.isFinite(n)) {
    // Python json.dumps emits NaN/Infinity/-Infinity as bare tokens; these
    // should never appear in a canon, but mirror the behavior rather than throw.
    if (Number.isNaN(n)) return "NaN";
    return n > 0 ? "Infinity" : "-Infinity";
  }
  return String(n);
}

/**
 * Serialize a value byte-identically to Python
 * `json.dumps(value, sort_keys=True, separators=(",",":"))` with the default
 * `ensure_ascii=True`. Objects sort keys by Unicode codepoint at every depth;
 * separators are compact; strings are ASCII-escaped; a `PyFloat` renders as a
 * Python float.
 */
function canonicalJson(value: unknown): string {
  if (value === null || value === undefined) return "null";
  if (value instanceof PyFloat) {
    const n = value.value;
    if (!Number.isFinite(n)) return encodeNumber(n);
    return Number.isInteger(n) ? `${n}.0` : String(n);
  }
  const t = typeof value;
  if (t === "boolean") return value ? "true" : "false";
  if (t === "number") return encodeNumber(value as number);
  if (t === "string") return encodeString(value as string);
  if (Array.isArray(value)) {
    return "[" + value.map((v) => canonicalJson(v)).join(",") + "]";
  }
  if (t === "object") {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).sort(); // JS default sort = Unicode codepoint
    const parts: string[] = [];
    for (const k of keys) {
      parts.push(encodeString(k) + ":" + canonicalJson(obj[k]));
    }
    return "{" + parts.join(",") + "}";
  }
  // Fallback mirrors Python's `default=str` (e.g. an unexpected type).
  return encodeString(String(value));
}

async function sha256Hex(input: string): Promise<string> {
  const bytes = new TextEncoder().encode(input);
  const digest = await crypto.subtle.digest("SHA-256", bytes);
  const view = new Uint8Array(digest);
  let hex = "";
  for (let i = 0; i < view.length; i++) hex += view[i]!.toString(16).padStart(2, "0");
  return hex;
}

/** Stable string form of a timestamp, matching Python `_normalize_dt`: a
 * `Date` -> `pyIsoformat` (Python `datetime.isoformat`), else `String(value)`. */
function normalizeDt(value: unknown): string {
  if (value instanceof Date) return pyIsoformat(value);
  return String(value);
}

/**
 * sha256 over an event's immutable content fields -- byte-identical to Python
 * `event_content_hash`. Excludes `eventId` and `entryHash` itself. `actor`/
 * `trust`/`recordedAt` are ALWAYS present (nullable); the claim-authority
 * fields are added ONLY when set (adding them as null would rehash every
 * provenance-free event and break existing seals).
 */
export async function eventContentHash(event: IdentityEvent): Promise<string> {
  const canon: Record<string, unknown> = {
    entity_id: event.entityId,
    kind: String(event.kind),
    payload: event.payload ?? null,
    run_name: event.runName ?? null,
    dataset: event.dataset ?? null,
    actor: event.actor ?? null,
    trust:
      event.trust !== undefined && event.trust !== null
        ? new PyFloat(round6(event.trust))
        : null,
    recorded_at: normalizeDt(event.recordedAt),
  };
  if (event.claimType !== undefined && event.claimType !== null) {
    canon["claim_type"] = String(event.claimType);
  }
  if (event.evidenceRef !== undefined && event.evidenceRef !== null) {
    canon["evidence_ref"] = String(event.evidenceRef);
  }
  if (event.previousClaimId !== undefined && event.previousClaimId !== null) {
    // Python `int(...)` -> integer repr (no decimal).
    canon["previous_claim_id"] = Math.trunc(event.previousClaimId);
  }
  const blob = canonicalJson(canon);
  return sha256Hex(blob);
}

/** The stored `entryHash` if present, else computed on the fly (covers rows
 * written before the hash column existed). Mirrors Python `_effective_hash`. */
async function effectiveHash(event: IdentityEvent): Promise<string> {
  return event.entryHash ?? (await eventContentHash(event));
}

/** One left-fold step: `acc' = sha256(acc || entryHash)`. Mirrors Python
 * `_fold_step`. A left fold means an incremental seal (seeded by the prior
 * root) yields the same root as folding the whole history from scratch. */
export async function foldStep(acc: string, entryHash: string): Promise<string> {
  return sha256Hex(acc + entryHash);
}

/** Result of `verifyAuditChain`. `ok` is the bottom line; the lists localize
 * what failed. Mirrors Python `AuditVerification`. */
export interface AuditVerification {
  ok: boolean;
  eventsChecked: number;
  sealsChecked: number;
  /** eventIds whose stored entryHash != recomputed content hash (content edit) */
  contentMismatches: number[];
  /** sealIds whose replayed root/count != stored (deletion/reorder/insertion) */
  sealMismatches: number[];
  /** sealIds whose lastEventId no longer exists in the log (sealed event deleted) */
  missingSealedEvents: number[];
}

export function verificationSummary(v: AuditVerification): string {
  if (v.ok) {
    return `audit chain intact: ${v.eventsChecked} events, ${v.sealsChecked} seals verified`;
  }
  const parts: string[] = [];
  if (v.contentMismatches.length) parts.push(`${v.contentMismatches.length} content edit(s)`);
  if (v.sealMismatches.length) parts.push(`${v.sealMismatches.length} seal mismatch(es)`);
  if (v.missingSealedEvents.length)
    parts.push(`${v.missingSealedEvents.length} sealed event(s) missing`);
  return "audit chain BROKEN: " + parts.join(", ");
}

/**
 * Anchor the current event log with a new seal and return it (or `null` when
 * there is nothing new to seal -- idempotent). Chains onto the latest seal for
 * `dataset` (`null` = global chain), folding only events appended since that
 * seal. Mirrors Python `seal_audit_log`.
 */
export async function sealAuditLog(
  store: IdentityStore,
  opts: { actor?: string | null; dataset?: string | null } = {},
): Promise<AuditSeal | null> {
  const dataset = opts.dataset ?? null;
  const actor = opts.actor ?? null;

  const prev = await store.latestSeal(dataset ?? undefined);
  const prevRoot = prev ? prev.rootHash : "";
  const prevLastId = prev && prev.lastEventId !== null ? prev.lastEventId : -1;
  const prevCount = prev ? prev.eventCount : 0;

  const all = await store.exportAuditLog(dataset ?? undefined);
  const newEvents = all.filter((e) => e.eventId !== null && e.eventId > prevLastId);
  if (newEvents.length === 0) return null;

  let acc = prevRoot;
  for (const e of newEvents) {
    acc = await foldStep(acc, await effectiveHash(e));
  }

  const seal: AuditSeal = {
    sealId: null,
    rootHash: acc,
    eventCount: prevCount + newEvents.length,
    lastEventId: newEvents[newEvents.length - 1]!.eventId,
    dataset,
    prevSealId: prev ? prev.sealId : null,
    prevRoot: prevRoot || null,
    actor,
    createdAt: new Date(0), // store assigns the real timestamp on write
  };
  const sealId = await store.addSeal(seal);
  seal.sealId = sealId;
  return seal;
}

/**
 * Replay the event log against its seal chain and report integrity. Two
 * independent checks (content edits; deletion/reorder/insertion of sealed
 * events). Mirrors Python `verify_audit_chain`.
 */
export async function verifyAuditChain(
  store: IdentityStore,
  opts: { dataset?: string | null } = {},
): Promise<AuditVerification> {
  const dataset = opts.dataset ?? null;
  const events = await store.exportAuditLog(dataset ?? undefined);
  const seals = await store.listSeals(dataset ?? undefined);

  const contentMismatches: number[] = [];
  for (const e of events) {
    if (e.entryHash !== undefined && e.entryHash !== null && e.eventId !== null) {
      if ((await eventContentHash(e)) !== e.entryHash) contentMismatches.push(e.eventId);
    }
  }

  const sealMismatches: number[] = [];
  const missingSealedEvents: number[] = [];
  const sealsSorted = [...seals].sort(
    (a, b) => (a.lastEventId ?? -1) - (b.lastEventId ?? -1),
  );
  let acc = "";
  let seen = 0;
  let sealIdx = 0;
  for (const e of events) {
    acc = await foldStep(acc, await effectiveHash(e));
    seen += 1;
    while (sealIdx < sealsSorted.length && sealsSorted[sealIdx]!.lastEventId === e.eventId) {
      const s = sealsSorted[sealIdx]!;
      if (acc !== s.rootHash || seen !== s.eventCount) {
        if (s.sealId !== null) sealMismatches.push(s.sealId);
      }
      sealIdx += 1;
    }
  }
  for (const s of sealsSorted.slice(sealIdx)) {
    if (s.sealId !== null) missingSealedEvents.push(s.sealId);
  }

  const ok =
    contentMismatches.length === 0 &&
    sealMismatches.length === 0 &&
    missingSealedEvents.length === 0;
  return {
    ok,
    eventsChecked: events.length,
    sealsChecked: seals.length,
    contentMismatches,
    sealMismatches,
    missingSealedEvents,
  };
}

File: packages/python/goldenmatch/goldenmatch/identity/audit.py
"""Tamper-evident audit log for the Identity Graph (#1078).

The identity event log is append-only, but "append-only by convention" is not
the same as "provably untampered". This module adds cryptographic
tamper-evidence in two contention-free layers:

1. **Per-event content hash** (``event_content_hash``) -- a sha256 over an
   event's own immutable fields, computed at insert time. It's a *pure*
   function (no prior-state read), so it adds no insert-time serialization
   point and works uniformly with the Postgres bulk-COPY write path. It catches
   any after-the-fact edit to a single event's content.

2. **On-demand seal chain** (``seal_audit_log`` / ``verify_audit_chain``) -- a
   periodic anchor that folds every event's content hash, in ``event_id``
   order, into a single chained root (git-/Certificate-Transparency-style).
   Each seal chains to its predecessor, so a verifier replaying the log can
   detect deletion, reordering, insertion, and content edits of any *sealed*
   event. Sealing is an explicit, infrequent operation -- never on the write
   hot path -- so the bulk/streaming ingest path is untouched.

Design note: tamper-evidence here is integrity, not secrecy. Anyone with write
access to the DB can also rewrite the seals; the value is that doing so
consistently across the event rows *and* the seal chain is detectable by an
external party who retains a prior seal root (publish it, mirror it, or store
it out-of-band). That is the same threat model as an append-only ledger.
"""
from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass
from datetime import datetime
from typing import TYPE_CHECKING, Any

from goldenmatch.identity.model import AuditSeal, IdentityEvent

if TYPE_CHECKING:
    from goldenmatch.identity.store import IdentityStore


def _normalize_dt(value: Any) -> str:
    """Stable string form of a timestamp, identical on the insert side (a
    ``datetime``) and the read-back side (``_to_dt`` returns a ``datetime``),
    so the content hash round-trips across a store write/read.

    **Dropping the tzinfo is load-bearing for the Postgres backend.** Events are
    created with a naive ``datetime.now()`` and hashed *before* insert, but the
    Postgres schema stores ``recorded_at`` as ``TIMESTAMPTZ`` and hands back a
    tz-AWARE datetime, whose ``isoformat()`` carries a ``+00:00`` offset the
    write-side hash never saw. Every stamped event therefore re-hashed to a
    different value and ``verify_audit_chain`` reported phantom "content
    edit(s)" on a perfectly clean in-DB store (caught by the `rust_pgrx`
    `gm_identity_audit_verify` smoke). Keeping the wall clock exactly as the
    driver returned it -- in the session timezone the naive value was written in
    -- makes the read-back string identical to the written one.

    Naive values (every SQLite event, and the write side on both backends) are
    untouched, so existing entry hashes and the seals chained over them stay
    valid -- the same back-compat constraint that governs the claim-authority
    fields in ``event_content_hash`` below.
    """
    if isinstance(value, datetime):
        if value.tzinfo is not None:
            value = value.replace(tzinfo=None)
        return value.isoformat()
    return str(value)


def event_content_hash(event: IdentityEvent) -> str:
    """sha256 over an event's immutable content fields (everything that
    identifies *what happened*, excluding the DB-assigned ``event_id`` and the
    ``entry_hash`` itself). Deterministic across SQLite/Postgres/Mongo and
    across a write->read round-trip: ``recorded_at`` is normalized to ISO-8601,
    ``trust`` rounded, ``payload`` canonicalized via sorted-key JSON."""
    canon = {
        "entity_id": event.entity_id,
        "kind": str(event.kind),
        "payload": event.payload,
        "run_name": event.run_name,
        "dataset": event.dataset,
        "actor": event.actor,
        "trust": round(event.trust, 6) if event.trust is not None else None,
        "recorded_at": _normalize_dt(event.recorded_at),
    }
    # Claim-authority fields (#1256) are hashed only when SET, not always-None
    # like actor/trust. Adding an always-present ``"claim_type": null`` key would
    # change every pre-#1256 event's hash and break existing seals; omitting the
    # keys when None keeps old (all-None) events byte-identical while still making
    # a set claim_type/evidence_ref/previous_claim_id tamper-evident (adding or
    # removing the key changes the blob).
    if event.claim_type is not None:
        canon["claim_type"] = str(event.claim_type)
    if event.evidence_ref is not None:
        canon["evidence_ref"] = str(event.evidence_ref)
    if event.previous_claim_id is not None:
        canon["previous_claim_id"] = int(event.previous_claim_id)
    blob = json.dumps(canon, sort_keys=True, separators=(",", ":"), default=str)
    return hashlib.sha256(blob.encode("utf-8")).hexdigest()


def _effective_hash(event: IdentityEvent) -> str:
    """The stored ``entry_hash`` if present, else computed on the fly. Lets the
    seal/verify path cover events written before the hash-chain column existed
    (migrated rows read back with ``entry_hash=None``)."""
    return event.entry_hash or event_content_hash(event)


def _fold_step(acc: str, entry_hash: str) -> str:
    """One left-fold step of the chain: ``acc' = sha256(acc || entry_hash)``.
    A left fold means an incremental seal (seeded by the prior root) yields the
    same root as folding the whole history from scratch."""
    return hashlib.sha256((acc + entry_hash).encode("utf-8")).hexdigest()


@dataclass
class AuditVerification:
    """Result of ``verify_audit_chain``. ``ok`` is the single bottom line; the
    lists localize *what* failed for an operator/report."""

    ok: bool
    events_checked: int
    seals_checked: int
    # event_ids whose stored entry_hash != recomputed content hash (content edit)
    content_mismatches: list[int]
    # seal_ids whose replayed root/count != stored (deletion/reorder/insertion)
    seal_mismatches: list[int]
    # seal_ids whose last_event_id no longer exists in the log (sealed event deleted)
    missing_sealed_events: list[int]

    def summary(self) -> str:
        if self.ok:
            return (
                f"audit chain intact: {self.events_checked} events, "
                f"{self.seals_checked} seals verified"
            )
        parts = []
        if self.content_mismatches:
            parts.append(f"{len(self.content_mismatches)} content edit(s)")
        if self.seal_mismatches:
            parts.append(f"{len(self.seal_mismatches)} seal mismatch(es)")
        if self.missing_sealed_events:
            parts.append(
                f"{len(self.missing_sealed_events)} sealed event(s) missing"
            )
        return "audit chain BROKEN: " + ", ".join(parts)

    def as_dict(self) -> dict[str, Any]:
        """JSON-ready verdict, the single serialization the MCP tool, the REST
        surface, and the SQL ``gm_identity_audit_verify`` bridge all emit."""
        return {
            "ok": self.ok,
            "events_checked": self.events_checked,
            "seals_checked": self.seals_checked,
            "content_mismatches": self.content_mismatches,
            "seal_mismatches": self.seal_mismatches,
            "missing_sealed_events": self.missing_sealed_events,
            "summary": self.summary(),
        }


def seal_result_dict(seal: AuditSeal | None) -> dict[str, Any]:
    """JSON-ready result for a ``seal_audit_log`` call. ``None`` (nothing new to
    seal) → ``{"sealed": False, ...}``; a fresh seal → its anchor fields. Single
    source for the MCP ``identity_audit_seal`` tool and the SQL
    ``gm_identity_audit_seal`` bridge."""
    if seal is None:
        return {"sealed": False, "reason": "no new events to seal"}
    return {
        "sealed": True,
        "seal_id": seal.seal_id,
        "root_hash": seal.root_hash,
        "event_count": seal.event_count,
        "last_event_id": seal.last_event_id,
        "dataset": seal.dataset,
        "actor": seal.actor,
    }


def seal_audit_log(
    store: IdentityStore,
    *,
    actor: str | None = None,
    dataset: str | None = None,
) -> AuditSeal | None:
    """Anchor the current event log with a new seal and return it.

    Chains onto the latest seal for ``dataset`` (``None`` = global chain),
    folding only the events appended since that seal so cost is proportional to
    new events, not the whole history. Returns ``None`` when there is nothing
    new to seal (idempotent: re-sealing an unchanged log is a no-op).
    """
    if getattr(store, "_backend", None) == "mongo":
        raise NotImplementedError(
            "seal_audit_log is not supported on the mongo backend"
        )
    prev = store.latest_seal(dataset=dataset)
    prev_root = prev.root_hash if prev else ""
    prev_last_id = prev.last_event_id if prev and prev.last_event_id is not None else -1
    prev_count = prev.event_count if prev else 0

    new_events = [
        e
        for e in store.export_audit_log(dataset=dataset)
        if e.event_id is not None and e.event_id > prev_last_id
    ]
    if not new_events:
        return None

    acc = prev_root
    for e in new_events:
        acc = _fold_step(acc, _effective_hash(e))

    seal = AuditSeal(
        root_hash=acc,
        event_count=prev_count + len(new_events),
        last_event_id=new_events[-1].event_id,
        dataset=dataset,
        prev_seal_id=prev.seal_id if prev else None,
        prev_root=prev_root or None,
        actor=actor,
    )
    seal_id = store.add_seal(seal)
    seal.seal_id = seal_id
    return seal


def verify_audit_chain(
    store: IdentityStore,
    *,
    dataset: str | None = None,
) -> AuditVerification:
    """Replay the event log against its seal chain and report integrity.

    Two independent checks:
      * **content** -- every event whose ``entry_hash`` was stored must still
        hash to it (catches an in-place edit of an event's fields);
      * **chain** -- replaying the fold over the *current* events must
        reproduce each seal's root and count at its boundary, and every sealed
        ``last_event_id`` must still exist (catches deletion, reordering, and
        insertion of sealed events).
    """
    if getattr(store, "_backend", None) == "mongo":
        raise NotImplementedError(
            "verify_audit_chain is not supported on the mongo backend"
        )
    events = store.export_audit_log(dataset=dataset)
    seals = store.list_seals(dataset=dataset)

    content_mismatches: list[int] = []
    for e in events:
        if e.entry_hash is not None and e.event_id is not None:
            if event_content_hash(e) != e.entry_hash:
                content_mismatches.append(e.event_id)

    # Single forward pass over events, checking each seal at its boundary.
    seal_mismatches: list[int] = []
    missing_sealed_events: list[int] = []
    seals_sorted = sorted(
        seals, key=lambda s: (s.last_event_id if s.last_event_id is not None else -1)
    )
    acc = ""
    seen = 0
    seal_idx = 0
    for e in events:
        acc = _fold_step(acc, _effective_hash(e))
        seen += 1
        while (
            seal_idx < len(seals_sorted)
            and seals_sorted[seal_idx].last_event_id == e.event_id
        ):
            s = seals_sorted[seal_idx]
            if acc != s.root_hash or seen != s.event_count:
                if s.seal_id is not None:
                    seal_mismatches.append(s.seal_id)
            seal_idx += 1
    # Any seal whose boundary event_id was never reached -> the sealed event was
    # deleted (or its id renumbered) out of the log.
    for s in seals_sorted[seal_idx:]:
        if s.seal_id is not None:
            missing_sealed_events.append(s.seal_id)

    ok = not (content_mismatches or seal_mismatches or missing_sealed_events)
    return AuditVerification(
        ok=ok,
        events_checked=len(events),
        seals_checked=len(seals),
        content_mismatches=content_mismatches,
        seal_mismatches=seal_mismatches,
        missing_sealed_events=missing_sealed_events,
    )


def audit_log_page(
    store: IdentityStore,
    *,
    dataset: str | None = None,
    actor: str | None = None,
    limit: int = 500,
) -> dict[str, Any]:
    """JSON-ready page of the append-only audit log for compliance export.

    Returns ``{"items": [...], "total": n}`` where ``items`` is truncated to
    ``limit`` (most recent-order preserved from ``export_audit_log``) and
    ``total`` is the full unfiltered-by-limit count. Single source for the MCP
    ``identity_audit`` tool and the SQL ``gm_identity_audit`` bridge."""
    events = store.export_audit_log(dataset=dataset, actor=actor)
    items = [
        {
            "event_id": e.event_id, "entity_id": e.entity_id, "kind": e.kind,
            "actor": e.actor, "trust": e.trust,
            "claim_type": e.claim_type, "evidence_ref": e.evidence_ref,
            "previous_claim_id": e.previous_claim_id,
            "recorded_at": e.recorded_at.isoformat() if e.recorded_at else None,
            "run_name": e.run_name, "dataset": e.dataset, "payload": e.payload,
        }
        for e in events[:limit]
    ]
    return {"items": items, "total": len(events)}
